ikarus/python/io/vtkwriter.hh

```diff
@@ -0,0 +1,99 @@
+ * This class supports adding result data as cell or point data and configuring VTK output formats.
+ *
+ * This function registers the following methods for the VtkWriter class:
+ * - `addAllResults(dataTag: DataTag)`
+ * - `addResult(resType: str, dataTag: ikarus.io.DataTag)`
+ * - `write(fileName)`, returns actual fileName
+ * - `addInterPolation(writer, vals_::np.array, basis, name: str, dataTag: ikarus.io.DataTag)`
+ * - `addPointData()` (multiple overloads)
+ * - `addCellData()` (multiple overloads)
+ * - `setFormat(type: dune.vtk.FormatTypes)`
+ * - `setDatatype(type: dune.vtk.DataTypes)`
+ * - `setHeadertype(type: dune.vtk.DataTypes)`
+ *
+ * \ingroup pythonbindings
+ *
+ * \tparam Writer The writer class type.
+ * \tparam options Additional options for the writer class.
+ * \param scope The scope in which to register the class.
+ * \param cls The class object to register the methods with.
+ */
+template <class Writer, class... options>
+void registerVtkWriter(pybind11::handle scope, pybind11::class_<Writer, options...> cls) {
+  using pybind11::operator""_a;
+
+  using Ikarus::Vtk::DataTag;
+
+  using Assembler     = typename Writer::Assembler;
+  using FE            = typename Writer::FEType;
+  using GridView      = typename Writer::GridView;
+  using VirtualizedGF = Dune::Vtk::Function<GridView>;
+
+  cls.def(pybind11::init(
+              [](std::shared_ptr<Assembler> assembler, Dune::Vtk::FormatTypes format, Dune::Vtk::DataTypes datatype,
+                 Dune::Vtk::DataTypes headertype) { return new Writer(assembler, format, datatype, headertype); }),
+          pybind11::arg("assembler"), pybind11::arg("format") = Dune::Vtk::FormatTypes::BINARY,
+          pybind11::arg("datatype")   = Dune::Vtk::DataTypes::FLOAT32,
+          pybind11::arg("headertype") = Dune::Vtk::DataTypes::UINT32, pybind11::keep_alive<1, 2>());
+
+  cls.def("setFormat", &Writer::setFormat);
+  cls.def("setDatatype", &Writer::setDatatype);
+  cls.def("setHeadertype", &Writer::setHeadertype);
+
+  cls.def(
+      "addAllResults", [](Writer& self, DataTag dataTag = Vtk::DataTag::asPointData) { self.addAllResults(dataTag); },
+      pybind11::arg("dataTag") = Vtk::DataTag::asPointData);
+
+  auto addResultImpl = [](Writer& self, const std::string& resType, auto type) {
+    bool success = false;
+    Dune::Hybrid::forEach(typename FE::SupportedResultTypes(), [&]<typename RT>(RT i) {
+      if (resType == toString(i)) {
+        success = true;
+        self.template addResult<RT::template Rebind>(type);
+      }
+    });
+    if (not success)
+      DUNE_THROW(Dune::NotImplemented, "Element " + Dune::className<FE>() + " doesn't support ResultType " + resType);
+  };
+
+  cls.def(
+      "addResult",
+      [&](Writer& self, const std::string& resType, DataTag tag = Vtk::DataTag::asPointData) {
+        addResultImpl(self, resType, tag);
+      },
+      pybind11::arg("resType"), pybind11::arg("tag") = Vtk::DataTag::asPointData);
+
+  cls.def(
```

Comment:
This function should not be needed since this is provided from the base class, isn't it?

Reply:
Yes, but this one returns the filename, which the base python method doens't do, but is the default C++ behaviour for dune-vtk

Reply:
But maybe we can also do a PR at dune-vtk for that

Reply:
Hmm but the `write` method has several overloads doesnt it? So we are only overwriting one method here. I'm also not sure how the pybind overload resolution chosen which function.
Also we are inheriting from a dune-vtk write doen't we? So we have trhe correct write function already from inheritance or?

Furthermore, I would suggest that I resolve the issues here otherwise I don't have an overview where I have to read your responses. Thanks!

Reply:
What i meant is that the python bindings and the c++ side for dune-vtk are not doing the same, i.e. returning the file name 
https://gitlab.dune-project.org/extensions/dune-vtk/-/blob/master/dune/vtk/vtkwriterbase.hh?ref_type=heads#L103
https://gitlab.dune-project.org/extensions/dune-vtk/-/blob/master/dune/python/vtk/writer.hh?ref_type=heads#L40

But yeah at least on python there is a second overload for the write function that we don't overwrite in our bindings here

Reply:
Ah ok then, I would suggest removing it here. If we need this we can create a small PR to `dune-vtk`

Reply:
Done